tests/heka/test_message_parser.py

```diff
@@ -1,87 +1,105 @@
-                    count += 1
-            except Exception:
-                threw = True
-        assert expected_exceptions[t] == threw
+    # enable this to regenerate the expected json representation of the ping
+    if False:
+        with open(filename, "rb") as f:
+            output = []
+            if "gzip" in heka_format:
+                f = streaming_gzip_wrapper(f)
+            for r in message_parser.parse_heka_message(f):
+                output.append(copy.deepcopy(r))
+            open(reference_filename, 'w').write(json.dumps(output, indent=4,
+                                                           sort_keys=True))
 
+    reference = json.load(open(reference_filename))
+    with open(filename, "rb") as f:
+        if "gzip" in heka_format:
+            f = streaming_gzip_wrapper(f)
+        for r in message_parser.parse_heka_message(f):
+            assert copy.deepcopy(r) == reference.pop()
 
-top_keys = set(["application", "clientId", "creationDate", "environment", "id", "meta",
-                "payload", "type", "version"])
-payload_keys = set(["UIMeasurements", "addonDetails", "childPayloads", "chromeHangs",
-                    "fileIOReports", "histograms", "info", "keyedHistograms", "lateWrites",
-                    "log", "processes", "simpleMeasurements", "slowSQL", "threadHangStats",
-                    "ver", "webrtc"])
-
-
-def test_telemetry(data_dir):
-    filename = "{}/test_telemetry_gzip.heka".format(data_dir)
-    with open(filename, "rb") as o:
-        for r in message_parser.parse_heka_message(streaming_gzip_wrapper(o)):
-            assert set(r.keys()) == top_keys
-            assert set(r["payload"].keys()) == payload_keys
-
-    filename = "{}/test_telemetry_snappy.heka".format(data_dir)
-    with open(filename, "rb") as o:
-        for r in message_parser.parse_heka_message(o):
-            assert set(r.keys()) == top_keys
-            assert set(r["payload"].keys()) == payload_keys
+    # should be no leftover reference data
+    assert len(reference) == 0
 
 
 def test_invalid_utf8(data_dir):
     filename = "{}/test_invalid_utf8.heka".format(data_dir)
     with open(filename, "rb") as o:
         for r in message_parser.parse_heka_message(o):
             assert(u'\ufffd' in r['info']['adapterDescription'])
+
+
+def test_lazy_parsing(data_dir, monkeypatch):
+    from moztelemetry.heka.message_parser import _parse_json
+    mock_parse_json = MagicMock(name='_parse_json',
+                                wraps=_parse_json)
+    import moztelemetry.heka.message_parser
+    monkeypatch.setattr(moztelemetry.heka.message_parser, '_parse_json', mock_parse_json)
+
+    filename = "{}/test_telemetry_gzip.heka".format(data_dir)
+    with open(filename, "rb") as o:
+        heka_message = message_parser.parse_heka_message(streaming_gzip_wrapper(o)).next()
+
+    # should only have parsed json *once* to get the payload (json/dictionary
+    # fields of the message should be parsed lazily)
```

Comment:
The comments here and below are helpful, but it might be nice to have a preliminary comment that explicitly mentions that the test input data has 20 fields, 19 of which should be parsed lazily. This can be inferred from the final assertion's comment, but it is not explicit.